<launch>
	<param name="use_sim_time" value="true"/>
	<include file="$(find cats)/launch/robot_state.launch" />
	<include file="$(find velodyne_pointcloud)/launch/32e_points.launch" />
	<include file="$(find velodyne_height_map)/launch/amsl_velodyne_heightmap.launch" />
	<node pkg="complement" type="complement" name="complement"/>
	<node pkg="rviz" type="rviz" name="rviz" args="-d /home/amsl/.rviz/road.rviz"/>
	<node pkg="rosbag" type="play" name="play" args="--clock /home/amsl/bagfiles/tsukuba_full_left_start.bag -s 2140 -l" />
   
	
	<group ns="recognition/road_recognizer">
        <arg name="hz" default="10"/>
        <arg name="normal_estimation_radius" default="0.60"/>
        <arg name="leaf_size" default="0.30"/>
        <!--arg name="curvature_threshold" default="0.005"/-->
        <arg name="curvature_threshold" default="0.90"/>
        <arg name="intensity_upper_threshold" default="255"/>
        <arg name="intensity_lower_threshold" default="20"/>
        <arg name="height_threshold" default="0.0"/>
        <arg name="max_random_sample_size" default="8000"/>
        <arg name="random_sample_ratio" default="0.20"/>

        <arg name="store_num" default="80"/>
        <arg name="position_difference_threshold" default="0.1"/>
        <arg name="yaw_difference_threshold" default="0.1"/>

        <arg name="outlier_removal_k" default="50"/>
        <arg name="outlier_removal_threshold" default="1.0"/>
        <arg name="enable_visualization" default="false"/>
        <arg name="beam_angle_num" default="720"/>
        <arg name="max_beam_range" default="20"/>
        <arg name="ransac_distance_threshold" default="0.20"/>
        <arg name="ransac_min_line_length_threshold" default="5.0"/>

		
		<node pkg="road_recognizer" type="make_image_node" name="make_image_node" output="screen">
			<remap from="/rm_ground" to="/cloud/static"/>
			<remap from="/grass" to="/recognition/road_recognizer/cloud/road/stored"/>
			<remap from="/image" to="/image/grass"/>
			<remap from="/image2" to="/image/edge"/>
            <param name="IMAGE_WIDTH" type="double" value="40"/>
			<param name="IMAGE_HEIGHT" type="double" value="40"/>
			<param name="IMAGE_RESILUTION" type="double" value="0.1"/>
			<param name="HOUGHLINE_FLAG" type="bool" value="false"/>
		</node>

        <node pkg="road_recognizer" type="road_cloud_publisher" name="road_cloud_publisher">
            <param name="HZ" value="$(arg hz)"/>
            <param name="NORMAL_ESTIMATION_RADIUS" value="$(arg normal_estimation_radius)"/>
            <param name="LEAF_SIZE" value="$(arg leaf_size)"/>
            <param name="CURVATURE_THRESHOLD" value="$(arg curvature_threshold)"/>
            <param name="INTENSITY_UPPER_THRESHOLD" value="$(arg intensity_upper_threshold)"/>
            <param name="INTENSITY_LOWER_THRESHOLD" value="$(arg intensity_lower_threshold)"/>
            <param name="HEIGHT_THRESHOLD" value="$(arg height_threshold)"/>
            <param name="MAX_RANDOM_SAMPLE_SIZE" value="$(arg max_random_sample_size)"/>
            <param name="RANDOM_SAMPLE_RATIO" value="$(arg random_sample_ratio)"/>
        </node>

        <node pkg="road_recognizer" type="road_point_cloud_storer" name="road_point_cloud_storer">
            <param name="HZ" value="$(arg hz)"/>
            <param name="STORE_NUM" value="$(arg store_num)"/>
            <param name="POSITION_DIFFERENCE_THRESHOLD" value="$(arg position_difference_threshold)"/>
            <param name="YAW_DIFFERENCE_THRESHOLD" value="$(arg yaw_difference_threshold)"/>
        </node>

        <node pkg="road_recognizer" type="road_recognizer" name="road_recognizer">
            <param name="HZ" value="$(arg hz)"/>
            <param name="LEAF_SIZE" value="$(arg leaf_size)"/>
            <param name="MAX_RANDOM_SAMPLE_SIZE" value="$(arg max_random_sample_size)"/>
            <param name="RANDOM_SAMPLE_RATIO" value="$(arg random_sample_ratio)"/>
            <param name="OUTLIER_REMOVAL_K" value="$(arg outlier_removal_k)"/>
            <param name="OUTLIER_REMOVAL_THRESHOLD" value="$(arg outlier_removal_threshold)"/>
            <param name="ENABLE_VISUALIZATION" value="$(arg enable_visualization)"/>
            <param name="BEAM_ANGLE_NUM" value="$(arg beam_angle_num)"/>
            <param name="MAX_BEAM_RANGE" value="$(arg max_beam_range)"/>
            <param name="RANSAC_DISTANCE_THRESHOLD" value="$(arg ransac_distance_threshold)"/>
            <param name="RANSAC_MIN_LINE_LENGTH_THRESHOLD" value="$(arg ransac_min_line_length_threshold)"/>
        </node>

    </group>

</launch>
